### ISSUE TITLE
bad values from Rice compressed FITS not handled correctly.

### ISSUE THREAD

**Issue by pdl-user** [PDLPorters/pdl#288]
When rfits reads a Rice compressed FITS file, pixels with the value BLANK are not marked as bad.
The code block below generates a 512x480 image with 256 32x30 blocks of values from -32768
to 32767 in steps of 257 and BLANK is set to -32768 . (Code for  gen_stepped_fits.pdl at bottom.) Then the CFITSIO utility imcopy is used to Rice compress and again to decompress.
```bash
$ ./gen_stepped_fits.pdl
$ imcopy stepped.fits stepped_rice.fits[compress]
$ imcopy stepped_rice.fits stepped_unrice0.fits
$ imcopy stepped_rice.fits[1] stepped_unrice.fits
$ pdl
perlDL shell v1.357
 PDL comes with ABSOLUTELY NO WARRANTY. For details, see the file
 'COPYING' in the PDL distribution. This is free software and you
 are welcome to redistribute it under certain conditions, see
 the same file for details.
ReadLines, NiceSlice, MultiLines  enabled
Reading .perldlrc...
Type 'demo' for online demos
Loaded PDL v2.024 (supports bad values)
pdl> $PDL::verbose = 1
pdl> $img = rfits 'stepped.fits'
Reading IMAGE data...
BITPIX =  16  size = 245760 pixels
Reading  491520  bytes
FITS file may contain bad values.
BSCALE =  &&  BZERO =

pdl> p $img->hdr->{BLANK}
-32768
pdl> p $img->min, $img->sum
-32511 31334400
```
So far, so good.

Next stepped_rice.fits is read with ds9 and the lower left corner block is properly displayed as BLANK (bad).
![image](https://user-images.githubusercontent.com/74427119/99737148-8d4ee000-2a7c-11eb-92c3-391bd2b50375.png)
But when stepped_rice.fits is read with PDL:
```
pdl> $img_rice = rfits 'stepped_rice.fits[1]'
Reading 3840 bytes of table data and 177600 bytes of heap data....

pdl> p $img_rice->hdr->{BLANK}
-32768
pdl> p $img_rice->min, $img_rice->sum
-32768 -122880
```
This time no print of:
`FITS file may contain bad values.`
and the minimum and sum are much lower.

When the Rice compressed file is decompressed, everything is OK again.

```
pdl> $img_unrice = rfits 'stepped_unrice.fits'
Reading IMAGE data...
BITPIX =  16  size = 245760 pixels
Reading  491520  bytes
FITS file may contain bad values.
BSCALE =  &&  BZERO =

pdl> p $img_unrice->hdr->{BLANK}
-32768
pdl> p $img_unrice->min, $img_unrice->sum
-32511 31334400
```
gen_stepped_fits.pdl source code:
```perl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::NiceSlice;
use PDL::IO::FITS;

print "$]\n";
my ($i, $j);
my $img = short zeroes(512, 480);
for ($j=0; $j<16; $j++) {
  for ($i=0; $i<16; $i++) {
    $img($i*32:$i*32+31, $j*30:$j*30+29) .= ($j*16 + $i)*257 - 32768;
  }
}
$img->hdr->{BLANK} = -32768;
$img->wfits('stepped.fits');
```

[fits_files.zip](https://github.com/PDLPorters/pdl/files/5591704/fits_files.zip)

**Comment by d-lamb** [comment]
Could you post stepped_rice.fits ?  My cfitsio from MacPorts doesn't come with imcopy and I'd rather not waste time  disentangling conflicting cfitsio libraries when I could be useful finding the actual bug here.

**Comment by pdl-user** [comment]
I would have put it in the original message, but github requires drag an drop and most of my systems don't have file browsers (which I loathe).

Dusted off one that does, got the files to it and attached them to the bottom of original post as a zip file.

**Comment by d-lamb** [comment]
Thanks for the zip file.  I think this bug is happening because the FITS header has no BSCALE keyword, not even the default of 1.0.  So when rfits is unpacking the tile-compressed image using the subroutine _rfits_unpack_zimage(), treat_bscale() is never called because BSCALE doesn't exist.  It is treat_bscale() that sets the BLANK value to BAD for each returned piddle.

If in your gen_stepped_fits.pdl you add a line before the wfits, `$img->hdr->{BSCALE}=1.0;`, and then generate the other files, does that solve the problem with PDL's reading of the rice-compressed file?  I'm hoping in this case that imcopy will preserve that BSCALE header field.  That may be a workaround, not a solution, but only if you control the generated FITS files in your real world use.

**Comment by pdl-user** [comment]
Thank you for looking into this.

Rather than changing gen_stepped_fits.pdl, I added BSCALE = 1.0 and BZERO = 0.0 into the header of a copy of a "real world" file and that worked. Unfortunately, the data set is over 10 years of over 50,000 4Kx4K images per day and growing.

**Comment by d-lamb** [comment]
I'm glad that worked, it confirmed what I thought was the problem.  Check out the small change I made on the linked PR and see if that fixes issue. I'd appreciate tests of:
- [ ] rfits of rice compressed with BLANK and BSCALE
- [ ] rfits of rice compressed with BLANK and no BSCALE
- [ ] rfits of rice compressed with no BLANK and BSCALE
- [ ] rfits of rice compressed with no BLANK and no BSCALE
- [ ] anything else relevant that you can think of.

It sounds like you're planning to read a bunch of SDO/AIA data?  Archive size, data rate, and image size is about right.  You might find that rfits is too slow.  Since you have cfitsio installed, it might be faster to directly interface with that library, perhaps through the [Astro::FITS::CFITSIO](https://metacpan.org/pod/Astro::FITS::CFITSIO) module.

**Comment by pdl-user** [comment]
Thank you for all your work on this.

I confirm that adding the line `$img->hdr->{BSCALE}=1.0;` to gen_stepped_fits.pdl then generating a Rice compressed file with imcopy results in a fits file that "current" rfits reads properly.

In addition BSCALE was set to 0.25 for a portion of the data a couple of years ago because of low data values. Current rfits also handles BAD values correctly for those files, so you are correct.

I know how to install PDL from my OS distribution and how to use perlbrew to build an arbitrary perl version and use cpanm to install PDL from CPAN, but do not know how to test bug fixes to PDL from a pull request.

I had forgotten that imcopy had been dropped from CFITSIO. I now recall that when I bumped into that, I copied imcopy.c from an older CFITSIO tinto the directory with the other C files and that `make imcopy` worked. I do not remember whether I had to tweak the Makefile, but I do know that imcopy.c ist standalone other than its library dependencies.

Is hope for `$img->wfits('stepped_rice.fits', {compress=>1});` working a lost cause?

In any case, for what it is worth, it seems to me that you have fixed the bug and that fix should be merged into the master branch and then this issue can be closed. 

Oh, and yes, SDO data and I have other means to work with the data, but sometimes PDL is or could be the best.
